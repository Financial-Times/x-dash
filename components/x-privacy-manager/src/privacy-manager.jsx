import { h } from '@financial-times/x-engine'

import { renderLoggedOutWarning, renderMessage } from './components/messages'
import { Form } from './components/form'
import { RadioBtn } from './components/radio-btn'
import { withCustomActions } from './actions'
import * as utils from './utils'

const defaultButtonText = {
	allow: { label: 'Allow', text: 'See personalised adverts' },
	block: { label: 'Block', text: 'Opt out of personalised adverts' },
	submit: { label: 'Save' }
}

/**
 * @param {XPrivacyManager.BasePrivacyManagerProps} Props
 */
export function BasePrivacyManager({
	userId,
	redirectUrl,
	legislationId,
	cookiesOnly,
	cookieDomain,
	fow,
	consent,
	consentSource,
	consentProxyApiHost,
	onConsentSavedCallbacks = [],
	buttonText = {},
	loginUrl,
	actions,
	isLoading,
	_response = undefined
}) {
	// Shallowly merge supplied button labels with defaults
	buttonText = { ...defaultButtonText, ...buttonText }

	const consentProxyEndpoints = utils.getConsentProxyEndpoints({
		userId,
		consentProxyApiHost,
		cookiesOnly,
		cookieDomain
	})
	const consentApiUrl = consentProxyEndpoints.createOrUpdateRecord
	const trackingKeys = utils.getTrackingKeys(legislationId)
	const { sendConsent, onConsentChange } = actions

	/**
	 * @param {"allow"|"block"} type
	 * @param {boolean} checked
	 */
	const radioBtnProps = (type, checked) => ({
		name: 'consent',
		type,
		checked,
		trackingKeys,
		buttonText,
		onChange: onConsentChange
	})

	/** @type {XPrivacyManager.FormProps} */
	const formProps = {
		consent,
		consentApiUrl,
		trackingKeys,
		buttonText,
		sendConsent: () => {
			return sendConsent({
				setConsentCookie: legislationId === 'gdpr',
				consentApiUrl,
				onConsentSavedCallbacks,
				consentSource,
				cookieDomain,
				fow
			})
		}
	}

	return (
		<div className="x-privacy-manager" data-component="x-privacy-manager">
			{renderLoggedOutWarning(userId, loginUrl)}
			<Form {...formProps}>
				<RadioBtn {...radioBtnProps('allow', consent === true)} />
				<RadioBtn {...radioBtnProps('block', consent === false)} type="block" checked={consent === false} />
			</Form>
			<div className="x-privacy-manager__messages" aria-live="polite">
				{renderMessage(isLoading, _response, redirectUrl)}
			</div>
		</div>
	)
}

export const PrivacyManager = withCustomActions(BasePrivacyManager)
