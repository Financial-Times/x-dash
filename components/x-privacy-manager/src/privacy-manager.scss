@import '@financial-times/o-buttons/main';
@import '@financial-times/o-colors/main';
@import '@financial-times/o-grid/main';
@import '@financial-times/o-spacing/main';
@import '@financial-times/o-typography/main';

@import './components/radio-btn';

%vertical-middle {
	display: inline-block;
	vertical-align: middle;
}

.x-privacy-manager__spinner {
	@extend %vertical-middle;
}

.x-privacy-manager__loading {
	@extend %vertical-middle;
	margin-left: oSpacingByName(s4);
}

.x-privacy-manager {
	@include oTypographySans($scale: 0, $line-height: 1.6);

	display: grid;
	gap: oSpacingByName(s8);

	& * {
		box-sizing: border-box;
	}
}

.x-privacy-manager__title {
	@include oTypographyHeading($level: 1);
}

.x-privacy-manager__consent-copy {
	padding: 0 oSpacingByName(m12);
	font-weight: 600;
	text-align: center;

	a {
		@include oTypographyLink;
	}
}

.x-privacy-manager-form {
	display: grid;
	gap: oSpacingByName(s8);
}

.x-privacy-manager-form__title {
	@include oTypographyHeading($level: 4);

	margin-top: oSpacingByName(s8);
}

.x-privacy-manager-form__controls {
	@include oGridRespondTo($from: S) {
		display: flex;
	}

	margin-top: oSpacingByName(s6);
}

.x-privacy-manager-form__submit {
	justify-self: center;

	@include oButtonsContent(
		$opts: (
			'type': 'primary',
			'size': 'big'
		)
	);

	display: block;
	padding: 0 oSpacingByName(m12);

	&:focus-visible {
		outline: 2px solid oColorsByName('teal-40');
		outline-offset: 3px;
		background-color: oColorsByName('teal-40');
	}
}
