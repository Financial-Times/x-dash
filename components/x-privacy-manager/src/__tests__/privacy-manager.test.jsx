const { h } = require('@financial-times/x-engine');
const { mount } = require('@financial-times/x-test-utils/enzyme');
const fetchMock = require('fetch-mock');

import { CONSENT_API, BasePrivacyManager, PrivacyManager } from '../privacy-manager';

fetchMock.mock(CONSENT_API, 200, { delay: 500 });

function checkPayload(opts, expected) {
	const values = Object.values(JSON.parse(String(opts.body)));
	return values.every((v) => v === expected);
}

describe('x-privacy-manager', () => {
	describe('initial state', () => {
		it('defaults to "Allow"', () => {
			const subject = mount(<PrivacyManager />);
			const inputTrue = subject.find('input[value="true"]').first();

			// Verify that initial props are correctly reflected
			expect(inputTrue.prop('checked')).toBe(true);
		});

		it('highlights explicitly set consent correctly: false', () => {
			const subject = mount(<PrivacyManager consent={false} />);
			const inputFalse = subject.find('input[value="false"]').first();

			// Verify that initial props are correctly reflected
			expect(inputFalse.prop('checked')).toBe(true);
		});

		it('highlights explicitly set consent correctly: true', () => {
			const subject = mount(<PrivacyManager consent={true} />);
			const inputTrue = subject.find('input[value="true"]').first();

			// Verify that initial props are correctly reflected
			expect(inputTrue.prop('checked')).toBe(true);
		});

		it('handles a change of consent', async () => {
			const subject = mount(<PrivacyManager consent={true} referrer="www.ft.com" />);
			const form = subject.find('form').first();
			const inputTrue = subject.find('input[value="true"]').first();
			const inputFalse = subject.find('input[value="false"]').first();
			const submitBtn = subject.find('button[type="submit"]');

			// Switch consent to false and submit form
			await inputFalse.prop('onChange')(undefined);
			await form.prop('onSubmit')(undefined);
			// Check that fetch was called with the correct values
			expect(checkPayload(fetchMock.lastOptions(), false)).toBe(true);

			// Switch consent back to true and resubmit form
			await inputTrue.prop('onChange')(undefined);
			await form.prop('onSubmit')(undefined);
			// Check that fetch was called with the correct values
			expect(checkPayload(fetchMock.lastOptions(), true)).toBe(true);

			// Reconcile snapshot with state
			subject.update();

			// Verify that confimatory nmessage is displayed
			const message = subject.find('[data-o-component="o-message"]').first();
			const link = message.find('[data-component="referrer-link"]');
			expect(message).toHaveClassName('o-message--success');
			expect(link).toHaveProp('href', 'https://www.ft.com/');
			expect(inputTrue).toHaveProp('checked', true);
			expect(submitBtn).toHaveProp('disabled', false);

			expect(subject).toMatchSnapshot();
		});
	});

	describe('It displays the appropriate messaging', () => {
		const defaultProps = {
			consent: true,
			referrer: 'www.ft.com',
			legislation: ['ccpa'],
			actions: {
				onConsentChange: jest.fn(() => {}),
				sendConsent: jest.fn().mockReturnValue({ _response: { ok: undefined } })
			},
			isLoading: false,
			_response: undefined
		};

		it('None by default', () => {
			const subject = mount(<BasePrivacyManager {...defaultProps} />);

			const messages = subject.find('[data-o-component="o-message"]');
			expect(messages).toHaveLength(0);

			expect(subject).toMatchSnapshot();
		});

		it('While loading', () => {
			const props = { ...defaultProps, isLoading: true };
			const subject = mount(<BasePrivacyManager {...props} />);

			const messages = subject.find('[data-o-component="o-message"]');
			expect(messages).toHaveLength(1);
			expect(messages.first()).toHaveClassName('o-message--neutral');

			const submitBtn = subject.find('button[type="submit"]');
			expect(submitBtn).toHaveProp('disabled', true);

			expect(subject).toMatchSnapshot();
		});

		it('On receiving a response with a status of 200', () => {
			const _response = { ok: true, status: 200 };
			const props = { ...defaultProps, _response };
			const subject = mount(<BasePrivacyManager {...props} />);

			const messages = subject.find('[data-o-component="o-message"]');
			const message = messages.first();

			expect(messages).toHaveLength(1);
			expect(message).toHaveClassName('o-message--success');

			const link = message.find('[data-component="referrer-link"]');
			expect(link).toHaveProp('href', 'https://www.ft.com/');

			const submitBtn = subject.find('button[type="submit"]');
			expect(submitBtn).toHaveProp('disabled', false);

			expect(subject).toMatchSnapshot();
		});

		it('On receiving a response with a non-200 status', () => {
			const _response = { ok: false, status: 400 };
			const props = { ...defaultProps, _response };
			const subject = mount(<BasePrivacyManager {...props} />);

			const messages = subject.find('[data-o-component="o-message"]');
			const message = messages.first();

			expect(messages).toHaveLength(1);
			expect(message).toHaveClassName('o-message--error');

			const link = message.find('[data-component="referrer-link"]');
			expect(link).toHaveProp('href', 'https://www.ft.com/');

			const submitBtn = subject.find('button[type="submit"]');
			expect(submitBtn).toHaveProp('disabled', false);

			expect(subject).toMatchSnapshot();
		});
	});
});
