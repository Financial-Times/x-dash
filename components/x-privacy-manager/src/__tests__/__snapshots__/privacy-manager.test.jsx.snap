// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`x-privacy-manager It displays the appropriate messaging None by default 1`] = `
<BasePrivacyManager
  actions={
    Object {
      "onConsentChange": [MockFunction],
      "sendConsent": [MockFunction],
    }
  }
  consent={true}
  isLoading={false}
  legislation={
    Array [
      "ccpa",
    ]
  }
  referrer="www.ft.com"
>
  <div>
    <h1>
      Do Not Sell My Personal Information
    </h1>
    <div>
      <p>
        CCPA defines the "sale" of personal information in extremely broad terms. For this reason, the collection of data by advertisers who advertise on our Sites for the purposes of measuring the effectiveness of their advertising is caught, and requires the ability to object to the use of your personal information. The data collected by advertisers may include online identifiers, such as IP address, and interactions with advertisements.
      </p>
      <p>
        Our advertising Terms and Conditions limit the use of this data by advertisers only to their own business purposes, including analytics and attribution, not for commercial purposes.
      </p>
      <hr />
      <form
        action="#"
        onSubmit={[Function]}
      >
        <h2>
          Use of my personal information for advertising purposes
        </h2>
        <div>
          <RadioBtn
            checked={true}
            onChange={[MockFunction]}
            value="true"
          >
            <div>
              <input
                checked={true}
                id="ccpa-true"
                name="consent"
                onChange={[MockFunction]}
                type="radio"
                value="true"
              />
              <label
                htmlFor="ccpa-true"
              >
                <span>
                  <strong>
                    Allow
                  </strong>
                  <span>
                    See personalised adverts
                  </span>
                </span>
                <svg
                  viewBox="0 0 36 36"
                >
                  <circle
                    cx="18"
                    cy="18"
                    r="16"
                  />
                  <circle
                    cx="18"
                    cy="18"
                    r="8"
                  />
                </svg>
              </label>
            </div>
          </RadioBtn>
          <RadioBtn
            checked={false}
            onChange={[MockFunction]}
            value="false"
          >
            <div>
              <input
                checked={false}
                id="ccpa-false"
                name="consent"
                onChange={[MockFunction]}
                type="radio"
                value="false"
              />
              <label
                htmlFor="ccpa-false"
              >
                <span>
                  <strong>
                    Block
                  </strong>
                  <span>
                    Only see generic adverts
                  </span>
                </span>
                <svg
                  viewBox="0 0 36 36"
                >
                  <circle
                    cx="18"
                    cy="18"
                    r="16"
                  />
                  <circle
                    cx="18"
                    cy="18"
                    r="8"
                  />
                </svg>
              </label>
            </div>
          </RadioBtn>
        </div>
        <button
          disabled={false}
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </div>
</BasePrivacyManager>
`;

exports[`x-privacy-manager It displays the appropriate messaging On receiving a response with a non-200 status 1`] = `
<BasePrivacyManager
  _response={
    Object {
      "ok": false,
      "status": 400,
    }
  }
  actions={
    Object {
      "onConsentChange": [MockFunction],
      "sendConsent": [MockFunction],
    }
  }
  consent={true}
  isLoading={false}
  legislation={
    Array [
      "ccpa",
    ]
  }
  referrer="www.ft.com"
>
  <div>
    <h1>
      Do Not Sell My Personal Information
    </h1>
    <div>
      <p>
        CCPA defines the "sale" of personal information in extremely broad terms. For this reason, the collection of data by advertisers who advertise on our Sites for the purposes of measuring the effectiveness of their advertising is caught, and requires the ability to object to the use of your personal information. The data collected by advertisers may include online identifiers, such as IP address, and interactions with advertisements.
      </p>
      <p>
        Our advertising Terms and Conditions limit the use of this data by advertisers only to their own business purposes, including analytics and attribution, not for commercial purposes.
      </p>
      <hr />
      <ResponseMessage
        referrer="www.ft.com"
        success={false}
      >
        <Message
          cls="o-message o-message--alert o-message--error"
        >
          <div
            className="o-message o-message--alert o-message--error"
            data-o-component="o-message"
          >
            <div
              className="o-message__container"
            >
              <div
                className="o-message__content "
              >
                <div
                  className="o-message__content-main"
                >
                  Your setting could not be saved. Please try again later.
                  Â 
                  <a
                    className="o-message__actions__secondary"
                    data-component="referrer-link"
                    href="https://www.ft.com/"
                  >
                    Continue to homepage
                  </a>
                </div>
              </div>
            </div>
          </div>
        </Message>
      </ResponseMessage>
      <form
        action="#"
        onSubmit={[Function]}
      >
        <h2>
          Use of my personal information for advertising purposes
        </h2>
        <div>
          <RadioBtn
            checked={true}
            onChange={[MockFunction]}
            value="true"
          >
            <div>
              <input
                checked={true}
                id="ccpa-true"
                name="consent"
                onChange={[MockFunction]}
                type="radio"
                value="true"
              />
              <label
                htmlFor="ccpa-true"
              >
                <span>
                  <strong>
                    Allow
                  </strong>
                  <span>
                    See personalised adverts
                  </span>
                </span>
                <svg
                  viewBox="0 0 36 36"
                >
                  <circle
                    cx="18"
                    cy="18"
                    r="16"
                  />
                  <circle
                    cx="18"
                    cy="18"
                    r="8"
                  />
                </svg>
              </label>
            </div>
          </RadioBtn>
          <RadioBtn
            checked={false}
            onChange={[MockFunction]}
            value="false"
          >
            <div>
              <input
                checked={false}
                id="ccpa-false"
                name="consent"
                onChange={[MockFunction]}
                type="radio"
                value="false"
              />
              <label
                htmlFor="ccpa-false"
              >
                <span>
                  <strong>
                    Block
                  </strong>
                  <span>
                    Only see generic adverts
                  </span>
                </span>
                <svg
                  viewBox="0 0 36 36"
                >
                  <circle
                    cx="18"
                    cy="18"
                    r="16"
                  />
                  <circle
                    cx="18"
                    cy="18"
                    r="8"
                  />
                </svg>
              </label>
            </div>
          </RadioBtn>
        </div>
        <button
          disabled={false}
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </div>
</BasePrivacyManager>
`;

exports[`x-privacy-manager It displays the appropriate messaging On receiving a response with a status of 200 1`] = `
<BasePrivacyManager
  _response={
    Object {
      "ok": true,
      "status": 200,
    }
  }
  actions={
    Object {
      "onConsentChange": [MockFunction],
      "sendConsent": [MockFunction],
    }
  }
  consent={true}
  isLoading={false}
  legislation={
    Array [
      "ccpa",
    ]
  }
  referrer="www.ft.com"
>
  <div>
    <h1>
      Do Not Sell My Personal Information
    </h1>
    <div>
      <p>
        CCPA defines the "sale" of personal information in extremely broad terms. For this reason, the collection of data by advertisers who advertise on our Sites for the purposes of measuring the effectiveness of their advertising is caught, and requires the ability to object to the use of your personal information. The data collected by advertisers may include online identifiers, such as IP address, and interactions with advertisements.
      </p>
      <p>
        Our advertising Terms and Conditions limit the use of this data by advertisers only to their own business purposes, including analytics and attribution, not for commercial purposes.
      </p>
      <hr />
      <ResponseMessage
        referrer="www.ft.com"
        success={true}
      >
        <Message
          cls="o-message o-message--alert o-message--success"
        >
          <div
            className="o-message o-message--alert o-message--success"
            data-o-component="o-message"
          >
            <div
              className="o-message__container"
            >
              <div
                className="o-message__content "
              >
                <div
                  className="o-message__content-main"
                >
                  Your setting has been saved.
                  Â 
                  <a
                    className="o-message__actions__secondary"
                    data-component="referrer-link"
                    href="https://www.ft.com/"
                  >
                    Continue to homepage
                  </a>
                </div>
              </div>
            </div>
          </div>
        </Message>
      </ResponseMessage>
      <form
        action="#"
        onSubmit={[Function]}
      >
        <h2>
          Use of my personal information for advertising purposes
        </h2>
        <div>
          <RadioBtn
            checked={true}
            onChange={[MockFunction]}
            value="true"
          >
            <div>
              <input
                checked={true}
                id="ccpa-true"
                name="consent"
                onChange={[MockFunction]}
                type="radio"
                value="true"
              />
              <label
                htmlFor="ccpa-true"
              >
                <span>
                  <strong>
                    Allow
                  </strong>
                  <span>
                    See personalised adverts
                  </span>
                </span>
                <svg
                  viewBox="0 0 36 36"
                >
                  <circle
                    cx="18"
                    cy="18"
                    r="16"
                  />
                  <circle
                    cx="18"
                    cy="18"
                    r="8"
                  />
                </svg>
              </label>
            </div>
          </RadioBtn>
          <RadioBtn
            checked={false}
            onChange={[MockFunction]}
            value="false"
          >
            <div>
              <input
                checked={false}
                id="ccpa-false"
                name="consent"
                onChange={[MockFunction]}
                type="radio"
                value="false"
              />
              <label
                htmlFor="ccpa-false"
              >
                <span>
                  <strong>
                    Block
                  </strong>
                  <span>
                    Only see generic adverts
                  </span>
                </span>
                <svg
                  viewBox="0 0 36 36"
                >
                  <circle
                    cx="18"
                    cy="18"
                    r="16"
                  />
                  <circle
                    cx="18"
                    cy="18"
                    r="8"
                  />
                </svg>
              </label>
            </div>
          </RadioBtn>
        </div>
        <button
          disabled={false}
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </div>
</BasePrivacyManager>
`;

exports[`x-privacy-manager It displays the appropriate messaging While loading 1`] = `
<BasePrivacyManager
  actions={
    Object {
      "onConsentChange": [MockFunction],
      "sendConsent": [MockFunction],
    }
  }
  consent={true}
  isLoading={true}
  legislation={
    Array [
      "ccpa",
    ]
  }
  referrer="www.ft.com"
>
  <div>
    <h1>
      Do Not Sell My Personal Information
    </h1>
    <div>
      <p>
        CCPA defines the "sale" of personal information in extremely broad terms. For this reason, the collection of data by advertisers who advertise on our Sites for the purposes of measuring the effectiveness of their advertising is caught, and requires the ability to object to the use of your personal information. The data collected by advertisers may include online identifiers, such as IP address, and interactions with advertisements.
      </p>
      <p>
        Our advertising Terms and Conditions limit the use of this data by advertisers only to their own business purposes, including analytics and attribution, not for commercial purposes.
      </p>
      <hr />
      <LoadingMessage>
        <Message
          cls="o-message o-message--neutral"
        >
          <div
            className="o-message o-message--neutral"
            data-o-component="o-message"
          >
            <div
              className="o-message__container"
            >
              <div
                className="o-message__content "
              >
                <div
                  className="o-message__content-main"
                >
                  <div
                    className="o-loading o-loading--dark o-loading--small undefined"
                  />
                  <span>
                    Loading...
                  </span>
                </div>
              </div>
            </div>
          </div>
        </Message>
      </LoadingMessage>
      <form
        action="#"
        onSubmit={[Function]}
      >
        <h2>
          Use of my personal information for advertising purposes
        </h2>
        <div>
          <RadioBtn
            checked={true}
            onChange={[MockFunction]}
            value="true"
          >
            <div>
              <input
                checked={true}
                id="ccpa-true"
                name="consent"
                onChange={[MockFunction]}
                type="radio"
                value="true"
              />
              <label
                htmlFor="ccpa-true"
              >
                <span>
                  <strong>
                    Allow
                  </strong>
                  <span>
                    See personalised adverts
                  </span>
                </span>
                <svg
                  viewBox="0 0 36 36"
                >
                  <circle
                    cx="18"
                    cy="18"
                    r="16"
                  />
                  <circle
                    cx="18"
                    cy="18"
                    r="8"
                  />
                </svg>
              </label>
            </div>
          </RadioBtn>
          <RadioBtn
            checked={false}
            onChange={[MockFunction]}
            value="false"
          >
            <div>
              <input
                checked={false}
                id="ccpa-false"
                name="consent"
                onChange={[MockFunction]}
                type="radio"
                value="false"
              />
              <label
                htmlFor="ccpa-false"
              >
                <span>
                  <strong>
                    Block
                  </strong>
                  <span>
                    Only see generic adverts
                  </span>
                </span>
                <svg
                  viewBox="0 0 36 36"
                >
                  <circle
                    cx="18"
                    cy="18"
                    r="16"
                  />
                  <circle
                    cx="18"
                    cy="18"
                    r="8"
                  />
                </svg>
              </label>
            </div>
          </RadioBtn>
        </div>
        <button
          disabled={true}
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </div>
</BasePrivacyManager>
`;

exports[`x-privacy-manager initial state handles a change of consent 1`] = `
<withActions(BasePrivacyManager)
  consent={true}
  referrer="www.ft.com"
>
  <InteractionClass
    Component={[Function]}
    actions={
      Object {
        "onConsentChange": [Function],
        "sendConsent": [Function],
      }
    }
    initialState={
      Object {
        "consent": true,
        "referrer": "www.ft.com",
      }
    }
  >
    <InteractionRender
      Component={[Function]}
      actions={
        Object {
          "onConsentChange": [Function],
          "sendConsent": [Function],
        }
      }
      inFlight={0}
      initialState={
        Object {
          "consent": true,
          "referrer": "www.ft.com",
        }
      }
      state={
        Object {
          "_response": Object {
            "ok": true,
          },
          "consent": true,
        }
      }
    >
      <BasePrivacyManager
        _response={
          Object {
            "ok": true,
          }
        }
        actions={
          Object {
            "onConsentChange": [Function],
            "sendConsent": [Function],
          }
        }
        consent={true}
        isLoading={false}
        referrer="www.ft.com"
      >
        <div>
          <h1>
            Do Not Sell My Personal Information
          </h1>
          <div>
            <p>
              CCPA defines the "sale" of personal information in extremely broad terms. For this reason, the collection of data by advertisers who advertise on our Sites for the purposes of measuring the effectiveness of their advertising is caught, and requires the ability to object to the use of your personal information. The data collected by advertisers may include online identifiers, such as IP address, and interactions with advertisements.
            </p>
            <p>
              Our advertising Terms and Conditions limit the use of this data by advertisers only to their own business purposes, including analytics and attribution, not for commercial purposes.
            </p>
            <hr />
            <ResponseMessage
              referrer="www.ft.com"
              success={true}
            >
              <Message
                cls="o-message o-message--alert o-message--success"
              >
                <div
                  className="o-message o-message--alert o-message--success"
                  data-o-component="o-message"
                >
                  <div
                    className="o-message__container"
                  >
                    <div
                      className="o-message__content "
                    >
                      <div
                        className="o-message__content-main"
                      >
                        Your setting has been saved.
                        Â 
                        <a
                          className="o-message__actions__secondary"
                          data-component="referrer-link"
                          href="https://www.ft.com/"
                        >
                          Continue to homepage
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </Message>
            </ResponseMessage>
            <form
              action="#"
              onSubmit={[Function]}
            >
              <h2>
                Use of my personal information for advertising purposes
              </h2>
              <div>
                <RadioBtn
                  checked={true}
                  onChange={[Function]}
                  value="true"
                >
                  <div>
                    <input
                      checked={true}
                      id="ccpa-true"
                      name="consent"
                      onChange={[Function]}
                      type="radio"
                      value="true"
                    />
                    <label
                      htmlFor="ccpa-true"
                    >
                      <span>
                        <strong>
                          Allow
                        </strong>
                        <span>
                          See personalised adverts
                        </span>
                      </span>
                      <svg
                        viewBox="0 0 36 36"
                      >
                        <circle
                          cx="18"
                          cy="18"
                          r="16"
                        />
                        <circle
                          cx="18"
                          cy="18"
                          r="8"
                        />
                      </svg>
                    </label>
                  </div>
                </RadioBtn>
                <RadioBtn
                  checked={false}
                  onChange={[Function]}
                  value="false"
                >
                  <div>
                    <input
                      checked={false}
                      id="ccpa-false"
                      name="consent"
                      onChange={[Function]}
                      type="radio"
                      value="false"
                    />
                    <label
                      htmlFor="ccpa-false"
                    >
                      <span>
                        <strong>
                          Block
                        </strong>
                        <span>
                          Only see generic adverts
                        </span>
                      </span>
                      <svg
                        viewBox="0 0 36 36"
                      >
                        <circle
                          cx="18"
                          cy="18"
                          r="16"
                        />
                        <circle
                          cx="18"
                          cy="18"
                          r="8"
                        />
                      </svg>
                    </label>
                  </div>
                </RadioBtn>
              </div>
              <button
                disabled={false}
                type="submit"
              >
                Save
              </button>
            </form>
          </div>
        </div>
      </BasePrivacyManager>
    </InteractionRender>
  </InteractionClass>
</withActions(BasePrivacyManager)>
`;
